### ISSUE TITLE
Question: ragged batch with ONNX backend

### ISSUE THREAD

**Issue by JulesBelveze** [triton-inference-server/server#3741]
Hi all,

I am posting it here as the _Triton Inference Server_ topic on the forum is archived.

I managed to run BERT model on a ONNX backend with the default scheduler and no batching algorithms and everything went smoothly.
 
I am now trying to use the ragged batch algorithm to fully leverage Triton. After reading the [ragged_batching documentation](https://github.com/triton-inference-server/server/blob/main/docs/ragged_batching.md) I've modified my configuration file to the following:
<details>
<summary>Click to expand</summary>

```
name: "bert"
platform: "onnxruntime_onnx"
max_batch_size: 256
input [
  {
    name: "input_ids"
    data_type: TYPE_INT64
    dims: [-1],
    allow_ragged_batch: true
  },
  {
    name: "token_type_ids"
    data_type: TYPE_INT64
    dims: [-1],
    allow_ragged_batch: true
  },
  {
    name: "attention_mask"
    data_type: TYPE_INT64
    dims: [-1],
    allow_ragged_batch: true
  }
]
batch_input [
  {
    kind: BATCH_ACCUMULATED_ELEMENT_COUNT
    target_name: "input_ids_index"
    data_type: TYPE_FP32
    source_input: "input_ids"
  },
  {
    kind: BATCH_ACCUMULATED_ELEMENT_COUNT
    target_name: "token_type_ids_index"
    data_type: TYPE_FP32
    source_input: "token_type_ids"
  },
  {
    kind: BATCH_ACCUMULATED_ELEMENT_COUNT
    target_name: "attention_mask_index"
    data_type: TYPE_FP32
    source_input: "attention_mask"
  }
]
output [
  {
    name: "start_logits"
    data_type: TYPE_FP32
    dims: [ -1, 5 ]
  },
  {
    name: "end_logits"
    data_type: TYPE_FP32
    dims: [-1, 5 ]
  }
]
dynamic_batching { }
```
</details>

but I am hitting this issue:
```
E1228 14:34:20.869115 1 model_repository_manager.cc:1186] failed to load 'bert' version 1: Invalid argument: unable to load model 'bert', configuration expects 6 inputs, model provides 3
```

I understand I need to modify my model to accept all parameters but I am not quite sure what changes do I need to perform on the model side.

I would really appreciate if someone could give me some hints to have it to work 😃 

Thanks in advance,
Cheers!

**Comment by dzier** [comment]
The batch input section tells Triton to dynamically generate those 3 additional input tensors, as defined, and pass them to the model for each request. The model must have these inputs. Additionally, the model must also be modified to handle these new inputs correctly. Since the inputs are ragged and variable in size, the input tensors will have different starting points. The index tensor will contain the index into the original vector for the input.

For example, in a ragged batch, 'input_ids' will be a 1-dim tensor and Triton will create 'input_ids_index' to indicate the index into 'input_ids' for each request in the batch. The model must be written to understand this. 

The existing BERT model cannot be expected to work with ragged batching out of the box.

The [BERT model](https://github.com/mlcommons/inference_results_v1.1/tree/main/closed/NVIDIA/configs/bert) used in NVIDIA's optimized inference MLPerf solution uses ragged batching on Triton can be used as a reference.

**Comment by JulesBelveze** [comment]
Hey @dzier thanks for the clarification! I'll check that out 🙂
